stream_color: Replace spectrum-colorpicker library with a new color picker popover.

sayamsamal opened the pull request:
This PR replaces the spectrum-colorpicker library with a new color picker popover which uses the HTML5 <input type="color"> element for native browser color picker for custom colors.

Fixes: #14961

### TODO:
- [x] **Fix keyboard navigation in the color-picker popover.**
_Currently the focus switches back to the stream actions popover._
- [ ] ~**Fix focus on the custom color picker/color preview area.**~ (The preview color rectangle has been removed )
~_Currently the focus seems to be lost while tabbing through it._~
- [X] **Debounce the function to set the custom color via the native browser color picker**
_This prevents API rate limits that can arise via multiple events when a user drags across the color spectrum map._
- [x] **Use radio buttons for the color swatches.**
_Radio buttons replicates the needed behavior: Only one color option can be selected at a time. This is also a step towards best practices as it also removes the need for custom keyboard navigation, since radio buttons already implement that functionality._

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Before | After - Light Mode | After - Dark Mode |
|--------|--------|--------|
| ![Screenshot 2024-12-12 at 2 05 33 AM](https://github.com/user-attachments/assets/501fc271-d7c6-4669-90e2-9c97ed40698c) | ![Screenshot 2024-12-12 at 1 54 49 AM](https://github.com/user-attachments/assets/67fdf360-5ba2-4d8f-961d-4474ef8bdaf8) | ![Screenshot 2024-12-12 at 1 55 21 AM](https://github.com/user-attachments/assets/2a582810-f435-4b4c-b12f-8e1c54ec56c6) | 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Cool! Playing with the PR:

1. Let's just preview the color with the channel name, without a `> topic`. That part just feels confusing, and is not needed.
2. The vertical alignment looks off here.
![Screenshot 2024-12-03 at 16 03 40](https://github.com/user-attachments/assets/9dfb1a3c-1fa0-459d-b998-09bd05863fc3)
3. "Apply" doesn't feel right. I expect an "Apply" button like that to keep the modal open, and just apply the change to the UI. I guess we could do it that way, but maybe we could have "Confirm" and "Cancel" buttons instead, as long as it's not too hard to implement?
4. As noted on CZO, I think, we need a way to actually pick a custom color and proceed from there. Probably adding Cancel/Confirm buttons there would be fine?

![Screenshot 2024-12-03 at 16 08 54](https://github.com/user-attachments/assets/e8ef71db-076a-4b17-9bc8-6fd698c8fe5d)

sayamsamal commented:
### Update
1. Removed the `> topic` part from the stream color preview located on the top of the color picker popover.
![Screenshot 2024-12-13 at 6 22 24 PM](https://github.com/user-attachments/assets/1c0984cd-8872-43fe-adc8-bbee8bdb49f7)
2. Fixed the vertical alignment of the "**Custom color**" popover option.
![Screenshot 2024-12-13 at 6 21 43 PM](https://github.com/user-attachments/assets/88222a63-5165-4210-840b-fa313ca906e2)
3. Removed the "**Apply**" button, and instead added debouncing to the setting of the custom color via the native browser color picker.
https://github.com/user-attachments/assets/114613cb-ea73-4e86-8c49-35ce3f791eb7
4.  Used radio buttons for the color swatches. Radio buttons replicates the needed behavior: Only one color option can be selected at a time. This is also a step towards best practices as it also removes the need for custom keyboard navigation, since radio buttons already implement that functionality.

sayamsamal commented:
@alya, could you give this PR another look?

alya commented:
I like it!

One little thing I noticed is that if you have "Custom color" selected with the keyboard, pressing Enter fails to open the custom color picker. Not a blocker to merging if we want to fix that as an immediate follow-up.

timabbott commented on web/src/color_picker_popover.ts:
```diff
@@ -0,0 +1,118 @@
+    },
+    200, // Wait for 200ms of inactivity
+    {leading: false}, // Don't execute immediately on the first color change
+);
```
We prefer to put comments on the line before the code it describes.

sayamsamal replied:
Moved the comments to the line before the code.

timabbott commented on web/templates/popovers/color_picker_popover.hbs:
```diff
@@ -0,0 +1,36 @@
+    </div>
+    <ul role="menu" class="popover-menu-list">
+        <li role="none">
+            <div role="group" class="color-swatch-list" aria-label="{{t 'App theme' }}">
```
Why is this "App theme" as the label for screenreaders?

sayamsamal replied:
Fixed. Sorry, I copied the theme switcher radio button group structure as the base for the color swatches, and missed these labels.

timabbott commented on web/templates/popovers/color_picker_popover.hbs:
```diff
@@ -0,0 +1,36 @@
+            <div role="group" class="color-swatch-list" aria-label="{{t 'App theme' }}">
+                {{#each stream_color_palette as |hex_color|}}
+                    <input type="radio" id="color-{{hex_color}}" class="color-swatch-input" name="color-picker-select" data-swatch-color="{{hex_color}}" {{#if (eq hex_color ../stream_color )}}checked{{/if}} />
+                    <label role="menuitemradio" class="color-swatch-label" for="color-{{hex_color}}" style="background-color: {{hex_color}};" aria-label="{{t 'Select automatic theme' }}" tabindex="0"></label>
```
This also looks copy-pasted from somewhere.

sayamsamal replied:
Fixed. Reason is the same as mentioned in https://github.com/zulip/zulip/pull/32063#discussion_r1893234532.

timabbott commented:
This looks pretty clean to me; my main request is that we should try having the stream popover hide when showing the color picker, since I think it's weird/distracting having several things open at once; maybe try doing that as a new commit stacked on top?

timabbott commented:
DM me when you have this updated; I'd like to get this integrated tomorrow if I can, to avoid it sitting unmerged over the holidays.

sayamsamal commented:
## Update
- Addressed the feedback from @timabbott
- Added a commit to hide the stream actions popover when opening the color picker popover. The color picker popover attaches to the same reference as that of the containing stream actions popover allowing us to visually replace the stream actions popover with the color picker popover.

timabbott commented:
Is this still missing the "OK" button proposed in chat.zulip.org? Or is that screencast out of date?

timabbott commented:
Merged as-is, and just noted in the design thread that we still need to add that. Thanks for all the work on this @sayamsamal!

sayamsamal commented:
Opened #32967 to incorporate the remaining feedback.
